webapp/src/components/llmbot_post.tsx

```diff
@@ -117,28 +136,70 @@ export const LLMBotPost = (props: Props) => {
 
+    // Get tool calls from post props
+    const toolCallsJson = props.post.props?.pending_tool_call;
+
+    // Update tool calls from props when available
     useEffect(() => {
-        const listenerID = Math.random().toString(36).substring(7);
-        props.websocketRegister(props.post.id, listenerID, (msg: WebSocketMessage<PostUpdateWebsocketMessage>) => {
-            const data = msg.data;
-            if (!data.control && !stoppedRef.current) {
-                setGenerating(true);
-                setMessage(data.next);
-            } else if (data.control === 'end') {
-                setGenerating(false);
-                setStopped(false);
-            } else if (data.control === 'start') {
-                setGenerating(true);
-                setStopped(false);
+        if (toolCallsJson) {
+            try {
+                const parsedToolCalls = JSON.parse(toolCallsJson);
+                setToolCalls(parsedToolCalls);
+            } catch (error) {
+                // Log error for debugging
+                setError('Error parsing tool calls');
             }
-        });
-        return () => {
-            props.websocketUnregister(props.post.id, listenerID);
-        };
-    }, []);
+        }
+    }, [toolCallsJson]);
+
+    useEffect(() => {
+        if (props.websocketRegister && props.websocketUnregister) {
+            const listenerID = Math.random().toString(36).substring(7);
+
+            props.websocketRegister(props.post.id, listenerID, (msg: WebSocketMessage<PostUpdateWebsocketMessage>) => {
+                const data = msg.data;
+
+                // Handle tool call events from the websocket event
+                if (data.control === 'tool_call' && data.post_id === props.post.id && data.tool_call) {
+                    try {
+                        const parsedToolCalls = JSON.parse(data.tool_call);
+                        setToolCalls(parsedToolCalls);
+                    } catch (error) {
+                        // Handle error silently
+                        setError('Error parsing tool call data');
+                    }
+                    return;
+                }
+
+                // Handle regular post updates
+                if (data.next && !stoppedRef.current) {
+                    setGenerating(true);
+                    setMessage(data.next);
+                } else if (data.control === 'end') {
+                    setGenerating(false);
+                    setStopped(false);
+                } else if (data.control === 'start') {
+                    setGenerating(true);
+                    setStopped(false);
+                }
+            });
+
+            return () => {
+                if (props.websocketUnregister) {
+                    props.websocketUnregister(props.post.id, listenerID);
+                }
+            };
+        }
+
+        return () => {/* no cleanup */};
+    }, [props.post.id]);
```

Comment:
I guess is not a problem, but should be the `props.websocketRegister`, and `props.websocketUnregister` part of the dependencies of this effect?

Reply:
I don't think it's an issue. I tend to not bother with the function dependencies and be explicit about what I think the useEffect should update on. 1/5

Reply:
Ok, for useEffect make sense